rpc/src/lib.rs

```diff
@@ -1,2 +1,192 @@
+
+    // Assemble
+    ///////////
+    Ok(ProverInput {
+        block_trace: BlockTrace {
+            trie_pre_images: pre_images.pop().context("trace had no BlockWitness")?,
+            txn_info,
+        },
+        other_data: OtherBlockData {
+            b_data: BlockLevelData {
+                b_meta: BlockMetadata {
+                    block_beneficiary: target_block.header.miner.compat(),
+                    block_timestamp: target_block.header.timestamp.into(),
+                    block_number: target_block_number.into(),
+                    block_difficulty: target_block.header.difficulty.into(),
+                    block_random: target_block
+                        .header
+                        .mix_hash
+                        .context("target block is missing field `mix_hash`")?
+                        .compat(),
+                    block_gaslimit: target_block.header.gas_limit.into(),
+                    block_chain_id: chain_id.into(),
+                    block_base_fee: target_block
+                        .header
+                        .base_fee_per_gas
+                        .context("target block is missing field `base_fee_per_gas`")?
+                        .into(),
+                    block_gas_used: target_block.header.gas_used.into(),
+                    block_bloom: target_block.header.logs_bloom.compat(),
+                },
+                b_hashes: BlockHashes {
+                    prev_hashes: prev_hashes.map(|it| it.compat()).into(),
+                    cur_hash: target_block
+                        .header
+                        .hash
+                        .context("target block is missing field `hash`")?
+                        .compat(),
+                },
+                withdrawals: target_block
+                    .withdrawals
+                    .into_iter()
+                    .flatten()
+                    .map(
+                        |Withdrawal {
+                             address, amount, ..
+                         }| { (address.compat(), amount.into()) },
+                    )
+                    .collect(),
+            },
+            checkpoint_state_trie_root: checkpoint_state_trie_root.compat(),
+        },
+    })
+}
+
+trait Compat<Out> {
+    fn compat(self) -> Out;
+}
+
+impl Compat<__compat_primitive_types::H160> for alloy::primitives::Address {
+    fn compat(self) -> __compat_primitive_types::H160 {
+        let alloy::primitives::Address(alloy::primitives::FixedBytes(arr)) = self;
+        __compat_primitive_types::H160(arr)
+    }
+}
+
+impl Compat<__compat_primitive_types::H256> for alloy::primitives::B256 {
+    fn compat(self) -> __compat_primitive_types::H256 {
+        let alloy::primitives::FixedBytes(arr) = self;
+        __compat_primitive_types::H256(arr)
+    }
+}
+
+impl Compat<[__compat_primitive_types::U256; 8]> for alloy::primitives::Bloom {
+    fn compat(self) -> [__compat_primitive_types::U256; 8] {
```

Comment:
Conversion could fail and crash, maybe we should introduce `try_compat`?

Reply:
This conversion cannot fail. The assert proves to a reader that there is no leftover data, and is covered in the test below

Reply:
I would argue that assert/panic should not be left in production software, although at the moment it is clear that conversion will succeed

Reply:
I disagree - `assert`s should maintain and document invariants, including in production code
- [here's the first such assert I found in the Rust standard library](https://github.com/rust-lang/rust/blob/7cf61ebde7b22796c69757901dd346d0fe70bd97/library/alloc/src/rc.rs#L1416)
- [here's the first such assert I found in tokio](https://github.com/tokio-rs/tokio/blob/91ad76c00c624ed607cf7fa53a6cc40f83f08159/tokio/src/fs/file.rs#L535)

Also worth noting that _any_ allocation can fail, triggering a runtime panic, and it's common for libraries to panic in other, documented, cases (e.g tokio futures polled outside of a tokio thread)